Nef_3/include/CGAL/Nef_3/SNC_decorator.h

```diff
@@ -824,8 +672,9 @@ class SNC_decorator : public SNC_const_decorator<Map> {
     return valid;
   }
 
-  template <typename Visitor>
-  void visit_shell_objects(SFace_handle f, Visitor& V) const;
+  template <typename Visitor, typename Traits = Self::Decorator_traits>
```

Comment:
Is this `typename Traits = Self::Decorator_traits` the best way to get the const or not const version of the decorator traits class in the loop? I don't see that we take anything else than this default.

Reply:
I can change this, but I am not sure of the syntax. Is the syntax:

```c++
template <typename Visitor, typename Traits>
void visit_shell_objects<Visitor,Self::Decorator_traits>(SFace_handle f, Visitor& V) const;
```

Reply:
@afabri Oh I remember why I did this now in SNC_const_decorator it uses Self::Decorator_traits AKA `SNC_const_decorator::Decorator_traits`, but in SNC_decorator it uses `SNC_decorator::Decorator_traits` so it does take something other than the default, despite the name looking idential.

Reply:
I think I just understood it.   In the call `Base::template visit_shell_objects<Visitor, Traits>(f, V);`  you call  the base class function but with the non-default traits.   So we might remove the `Traits`  in this overload in the non-const derived class, as for this overload we will never take something else.

Reply:
Ok, so I think I could change it like this:

```diff
diff --git a/Nef_3/include/CGAL/Nef_3/SNC_decorator.h b/Nef_3/include/CGAL/Nef_3/SNC_decorator.h
index 1074672824..791cbdd7d6 100644
--- a/Nef_3/include/CGAL/Nef_3/SNC_decorator.h
+++ b/Nef_3/include/CGAL/Nef_3/SNC_decorator.h
@@ -342,20 +342,17 @@ class SNC_decorator : public SNC_const_decorator<Map> {
     return( e1!=e2 && SD.cyclic_adj_succ(e2)==e1);
   }
 
-  template <typename Traits = Self::Decorator_traits>
-  typename Traits::Halffacet_handle get_visible_facet( const typename Traits::Vertex_handle v,
+  typename Self::Decorator_traits::Halffacet_handle get_visible_facet( const typename Self::Decorator_traits::Vertex_handle v,
                                                        const Ray_3& ray) const
-  { return Base::template get_visible_facet<Traits>(v, ray); }
+  { return Base::template get_visible_facet<typename Self::Decorator_traits>(v, ray); }
 
-  template <typename Traits = Self::Decorator_traits>
-  typename Traits::Halffacet_handle get_visible_facet( const typename Traits::Halfedge_handle e,
+  typename Self::Decorator_traits::Halffacet_handle get_visible_facet( const typename Self::Decorator_traits::Halfedge_handle e,
                                                        const Ray_3& ray) const
-  { return Base::template get_visible_facet<Traits>(e, ray); }
+  { return Base::template get_visible_facet<typename Self::Decorator_traits>(e, ray); }
 
-  template <typename Traits = Self::Decorator_traits>
-  typename Traits::Halffacet_handle get_visible_facet( const typename Traits::Halffacet_handle f,
+  typename Self::Decorator_traits::Halffacet_handle get_visible_facet( const typename Self::Decorator_traits::Halffacet_handle f,
                                                        const Ray_3& ray) const
-  { return Base::template get_visible_facet<Traits>(f, ray); }
+  { return Base::template get_visible_facet<typename Self::Decorator_traits>(f, ray); }
 
   Halffacet_handle get_visible_facet( const Vertex_handle v,
                                       const Segment_3& ray) const
@@ -672,9 +669,9 @@ class SNC_decorator : public SNC_const_decorator<Map> {
     return valid;
   }
 
-  template <typename Visitor, typename Traits = Self::Decorator_traits>
-  void visit_shell_objects(typename Traits::SFace_handle f, Visitor& V) const
-  { Base::template visit_shell_objects<Visitor, Traits>(f, V); }
+  template <typename Visitor>
+  void visit_shell_objects(typename Self::Decorator_traits::SFace_handle f, Visitor& V) const
+  { Base::template visit_shell_objects<Visitor,typename Self::Decorator_traits>(f, V); }
```

I do prefer it the way it is.

Reply:
Maybe I am wrong, but instead of `return Base::template get_visible_facet<typename Self::Decorator_traits>(e, ray);` can't  you just write `return Base::template get_visible_facet<Decorator_traits>(e, ray);`?
And as the above functions are in the derived class, instead of `typename Traits::Halffacet_handle` just `Halffacet_handle`  should do it.

Reply:
@afabri That's simpler ill give it a try. Thanks!

Reply:
If you do it quickly it might make it  into the testsuite with 5.3 in mind.  You see that I work on the backlog.